### PR TITLE
get organization members from GitHub

### PR DESCRIPTION
**Pull request by ghost**
Piggy-back allowing travis-ci to build all brances (didn't notice that
restriction earlier).

Purposely left out the following, they make for small introductory
tasks:
– using a library to parse and document command-line args
– making a GitHub.User object and capturing more fields from the API
– proper logging instead of print

Also for the future:
– discover the URLs from GitHub's API and find a library to parse those
templates with optional arguments, e.g.:
{
"organization_repositories_url":
"https://api.github.com/orgs/{org}/repos{?type,page,per_page,sort}",
…
}

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -11,10 +11,6 @@ matrix:
   allow_failures:
    - env: GHCVER=head CABALVER=head
 
-branches:
-  only:
-    - master
```
Comment (phadej) [inline .travis.yml:16]: this was here so we don't build twice, the branch and PR:
- https://travis-ci.org/futurice/fum2github/builds/61325572
- https://travis-ci.org/futurice/fum2github/builds/61326371

IMHO totally unnecessary to build feature branches

**Comment by phadej** [comment]
How nitpicky haskell-related comments do you want?

File: .travis.yml (hunk 1 above)
Comment (ghost) [inline .travis.yml:16]: I like the extra confirmation that tests pass on a clean build, even before making a PR.

**Comment by ghost** [comment]
Very nitpicky please. I'm using simple constructs (anything past Applicative Functors I wouldn't understand) but otherwise want thorough reviews.

File: library/Fum2GitHub/GitHub.hs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    responseBody,
+    responseHeaders,
+    withManager)
+import           Network.HTTP.Types.Header (ResponseHeaders)
+import           Text.Regex (mkRegex, matchRegex)
+
+
+-- Get the response body and headers from url using oAuthToken.
+getHttp :: String -> String -> IO (LBS.ByteString, ResponseHeaders)
+getHttp url oAuthToken = do
```
Comment (phadej) [inline library/Fum2GitHub/GitHub.hs:38]: this seems to be very same as in other module. Refactor into own module?

File: library/Fum2GitHub/GitHub.hs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    putStrLn url -- debug logging
+    baseReq <- parseUrl url
+    let agHdr = ("User-Agent", "https://github.com/futurice/fum2github")
+        authReq = applyBasicAuth (E.encodeUtf8 . T.pack $ oAuthToken)
+          "x-oauth-basic" baseReq
```
Comment (phadej) [inline library/Fum2GitHub/GitHub.hs]: this line wrapping is a bit distrubing

File: library/Fum2GitHub/GitHub.hs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    requestHeaders,
+    responseBody,
+    responseHeaders,
+    withManager)
+import           Network.HTTP.Types.Header (ResponseHeaders)
+import           Text.Regex (mkRegex, matchRegex)
+
+
+-- Get the response body and headers from url using oAuthToken.
+getHttp :: String -> String -> IO (LBS.ByteString, ResponseHeaders)
```
Comment (phadej) [inline library/Fum2GitHub/GitHub.hs]: would add `type URL = String` and `type OAuthToken = String` for better documentation.
And using `newtype` would prevent to accidentally swapping the params.

File: library/Fum2GitHub/GitHub.hs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        case getNextUrl hdrs of
+          Nothing -> return $ Right arr
+          Just nextUrl -> do
+            tailE <- getAll nextUrl oAuthToken
+            case tailE of
+              Left msg -> return $ Left msg
+              Right tail -> return $ Right (arr ++ tail)
+
+
+-- Get all members of the organisation.
```
Comment (phadej) [inline library/Fum2GitHub/GitHub.hs:97]: would create

``` hs
newtype GithubMember = GithubMember { getGithubMember :: String }
  deriving (Eq, Ord, Show)
```

and write a `FromJSON GithubMember` instance.

Using that, lots of boilerplate disappears.

File: library/Fum2GitHub/GitHub.hs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+import           Text.Regex (mkRegex, matchRegex)
+
+
+newtype URL = URL { getURL :: String } deriving (Eq, Show)
+
+newtype OAuthToken = OAuthToken { getOAuthToken :: String }
+
+
+-- Get the response body and headers from url using oAuthToken.
+getHttp :: URL -> OAuthToken -> IO (LBS.ByteString, ResponseHeaders)
```
Comment (ghost) [inline library/Fum2GitHub/GitHub.hs:37]: The `getHttp`s in Fum.hs and GitHub.hs are a bit different now (they add different headers and one also returns the ResponseHeaders).

But this suggested a bigger refactor, in a separate PR, which puts `getHttp` and traversing all API pages in a common module:

``` hs
-- Util.hs
getHttp :: Request -> Response

getAllJSON :: Request -> (Response -> Maybe Request) -> (Response -> Either String [a]) -> IO Either String [a]
getAllJSON initialReq getNextReq getResults =
  concatenate the results from all API pages
```

Fum/GitHub have different code for getting the next URL (http headers vs json body) and JSON results (different paths in the response body) but duplicate the logic to get the next page of results and concatenate them.

Unless such a refactor would over-complicate things?

**Pull request merged**